src/Message.php

```diff
@@ -405,30 +411,189 @@ public function clearAttachments()
 
   /**
    * Send the message
-   *
+   * We try to attempt to send the message MAX_RETRY_ATTEMPTS times. If it fails,
+   * we will push the message into the queue. Another implementation could be
+   * a blocking implementation where the number of retries and the wait timeout could
+   * be specified through the method call. We are not going ahead with that at this
+   * point since it might end up blocking. We are also not waiting before retry for
+   * the same reason.
    * @param string $text The text to send
    * @return void
    */
-  public function send($text = null)
+  protected function _send($text = null, $numRetries = self::MAX_RETRY_ATTEMPTS)
   {
-    if ($text) $this->setText($text);
+    if($text)
+    {
+      $this->setText($text);
+    }
+
+    $isMessageSent = false;
+
+    $numAttempts = 0;
+
+    // check for malicious calls and if so, try once at least
+    if ($numRetries <= 0)
+    {
+      $numRetries = 1;
+    }
+
+    while(($numAttempts < $numRetries) and
+          ($isMessageSent === false))
+    {
+      try
+      {
+        $this->client->sendMessage($this);
+
+        $isMessageSent = true;
+      }
+      catch(Exception $e){}
+
+      $numAttempts += 1;
+    }
 
-    $this->client->sendMessage($this);
+    // push it into the queue
+    if(!$isMessageSent)
+    {
+      $this->queue($text, $numRetries);
+    }
   }
 
   /**
    * Queue the message
    * @param  string $text The text to send
+   * @param  integer $numRetries The number of times to retry on failure
    * @return void
    */
-  public function queue($text = null)
+  protected function _queue($text = null, $numRetries)
   {
-    if ($text) $this->setText($text);
+    if ($text)
+    {
+        $this->setText($text);
+    }
 
-    $this->client->queueMessage($this, $this->queue);
+    $this->client->queueMessage($this, $this->queue, $numRetries);
```

Comment:
The method call here seems to be wrong. Client's `queueMessage` accepts only two parameters.